jsparagus/aps.py

```diff
@@ -190,44 +209,62 @@ def shift_next(self, pt: ParseTable) -> typing.Iterator[APS]:
         st, sh, la, rp, hs = self.stack, self.shift, self.lookahead, self.replay, self.history
         last_edge = sh[-1]
         state = pt.states[last_edge.src]
-        state_match_shift_end = True
+        state_match_shift_end = self.state == self.shift[-1].src
         if self.replay == []:
+            assert state_match_shift_end
             for term, to in state.shifted_edges():
-                edge = Edge(last_edge.src, term)
+                edge = Edge(self.state, term)
                 new_sh = self.shift[:-1] + [edge]
                 edge_to = Edge(to, None)
-                yield APS(st, new_sh + [edge_to], la + [term], rp, hs + [edge], False)
-        else:
+                yield APS(to, st, new_sh + [edge_to], la + [term], rp, hs + [edge], False)
+        elif state_match_shift_end:
             term = self.replay[0]
             rp = self.replay[1:]
             if term in state:
-                edge = Edge(last_edge.src, term)
+                edge = Edge(self.state, term)
                 new_sh = self.shift[:-1] + [edge]
                 to = state[term]
                 edge_to = Edge(to, None)
-                yield APS(st, new_sh + [edge_to], la, rp, hs + [edge], False)
+                yield APS(to, st, new_sh + [edge_to], la, rp, hs + [edge], False)
 
         rp = self.replay
         for a, to in state.epsilon:
-            edge = Edge(last_edge.src, a)
+            edge = Edge(self.state, a)
             prev_sh = self.shift[:-1] + [edge]
             # TODO: Add support for Lookahead and flag manipulation rules, as
             # both of these would invalide potential reduce paths.
             if a.update_stack():
+                stack_diff = a.update_stack_with()
+                if stack_diff.replay < 0:
+                    assert stack_diff.pop == 0
+                    assert stack_diff.nt is None
+                    num_replay = -stack_diff.replay
+                    assert len(self.replay) >= num_replay
+                    new_rp = self.replay[:]
+                    new_sh = self.shift[:]
+                    while num_replay > 0:
```

Comment:
A Replay action that replays multiple symbols is equivalent to a chain of Replay actions that each replay a single symbol, right? What if we did it that way? It seems simpler.

I am wary of this loop, partly because `shift_next` is documented to work one step at a time. Callers could use it to determine, for example, which states are reachable. Skipping states might be bad.

But mainly because the simpler thing seems like it would work.

Reply:
> A Replay action that replays multiple symbols is equivalent to a chain of Replay actions that each replay a single symbol, right? What if we did it that way? It seems simpler.

That's how I did it at first, them I moved to a list, as this reduces the number of nodes from the parse table, the number of action functions in the backend, and the (inacceptable) time spent iterating over these nodes.

> I am wary of this loop, partly because `shift_next` is documented to work one step at a time. Callers could use it to determine, for example, which states are reachable. Skipping states might be bad.

No, I do not think this is in anyway bad, as this is literally an optimization of a path which is already encoded in the parse table. The single step to be executed is `Replay`, which manipulates the parser state at once. Would `Reduce` pop each element one at a time be a saner approach? I do not think this matters.